### PR TITLE
cmd/podman/utils.go: Cancel-able resize writes

### PR DESCRIPTION
**Pull request by wking**
On Thu, Jun 28, 2018 at 03:48:26AM -0700, Marco Vedovati [wrote][1]:
> The root cause is a deadlock between two channel writes made by two different goroutines:
>
> 1. `resizeTty() : go func(){} : sendUpdate()` is sending a resize  message thru `resize` right at the beginning, but the channel is never read if some startup error occurs.
>
> 2. Upon program termination, `startAttachCtr() : defer func(){} ` is telling the goroutine in "1." to stop via the `resizeTerminate` channel. But that guy is still waiting for the write to `resize` to complete so the the termination message is never read.
>
> I think the go deadlock detection does not kick in because not all goroutines are seen as asleep. E.g. `os/signal Notify()` is enough to have the deadlock not detected.

333ab8c2 (#1010) addressed this with a deferred drain.  This commit adjusts that approach to use a single select tocover "have we been canceled?", "has there been a resize signal?", and (when we have one) "can we write the most recent resize event to the resize channel?".

A side benefit to this approach is that if we have a slow `resize` consumer and several resize signals, the `resizeTty` function will keep updating its local `resizeEvent`.  Once the `resize` channel is able to accept, only the most-recent event will be written.  Previously we'd have written one resize event for every received signal, even if the `resize` consumer was falling behind.

CC @marcov.

[1]: https://github.com/projectatomic/libpod/pull/1010#issuecomment-400994436

### CODE REVIEW

**Comment by rhatdan** [comment]
@baude @mheon @giuseppe PTAL

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

File: cmd/podman/utils.go [hunk 1]
```diff
@@ -171,33 +144,45 @@ func startAttachCtr(ctr *libpod.Container, stdout, stderr, stdin *os.File, detac
...
+					resizeEvent = getResize()
+				}
+			} else {
+				select {
+				case <-ctx.Done():
+					return
+				case <-sigchan:
+					resizeEvent = getResize()
+				case resize <- *resizeEvent:
+				}
```
Comment (giuseppe) [inline cmd/podman/utils.go]: shouldn't we reset resizeEvent here to not have an endless loop?

**Comment by giuseppe** [comment]
Please take a look at: https://github.com/projectatomic/libpod/pull/1020

**Pull request closed without merging**